### PR TITLE
Add WallpaperDownloader

### PR DESCRIPTION
**Pull request by guihkx**
About WallpaperDownloader:

>This is a JAVA software project developed for downloading, managing and changing your favorite wallpapers from different sources in Internet via GUI. It is a crossed-platform standalone application: it runs in Linux (GNOME Shell, KDE Plasma 5.8 (and higher), Unity, MATE, XFCE, Cinnamon, Pantheon and Budgie DE supported), Microsoft Windows (7,8 and 10) and macOS.

Source: https://bitbucket.org/eloy_garcia_pca/wallpaperdownloader

---

The developer and maintainer of WallpaperDownloader is @egara. I just created this Flatpak manifest. So, if the app gets accepted, please add both of us as maintainers.

Unfortunately, WallpaperDownloader currently makes use of `flatpak-spawn --host` to [run programs outside of the Flatpak sandbox](https://bitbucket.org/search?q=repo%3Awallpaperdownloader%20execProgram&account=%7B73a964e2-f01d-4860-b651-cff68acf2a28%7D).

As far as I know, since one of the premisses of WallpaperDownloader is to set wallpapers in the host, there's currently no way to around that.

FWIW, I've tried using the `--talk-name=ca.desrt.dconf` permission, but for some reason I'm not aware of, some schemas just weren't available (e.g. `org.mate.background` from MATE desktop).

Long story short: I thought it would be easier to just patch WallpaperDownloader to use `flatpak-spawn --host` to call `gsettings` or `xfconf-query` whenever inside the  Flatpak sandbox. I know that's not exactly good practice, but it's the only one I could think of. I'm open to suggestions, though!

So, since WallpaperDownloader can leave the Flatpak sandbox, I thought it would add an extra layer of transparency if it was built from source, instead of grabbing the pre-compiled `.jar` binary. That's why the `maven-dependencies.yml` file is there, because Flathub disallows network access during the build phase, therefore maven can't download dependencies. I explain how I generated that file [here](https://bitbucket.org/guihkx/wallpaperdownloader/src/flatpak/flatpak/README.md#markdown-header-generating-maven-dependenciesyml).

That's it, I guess.
Thanks!

### CODE REVIEW

**Comment by guihkx** [comment]
bot, build es.estoes.wallpaperDownloader

**Comment by guihkx** [comment]
bot, build es.estoes.wallpaperDownloader

**Comment by guihkx** [comment]
The screenshots are temporarily hosted on Imgur. I'll upload them to the Flathub repo if the app is accepted.

**Comment by guihkx** [comment]
bot, build es.estoes.wallpaperDownloader

**Comment by barthalion** [comment]
> As far as I know, since one of the premisses of WallpaperDownloader is to set wallpapers in the host, there's currently no way to around that.

xdg-desktop-portal implements wallpaper portal but you would have to sort out how to use it from Java. I'm not sure if any C app already uses it. Any ideas @matthiasclasen?

File: es.estoes.wallpaperDownloader.svg [hunk 1]
```diff
@@ -0,0 +1,101 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (barthalion) [inline es.estoes.wallpaperDownloader.svg]: Ideally the icon, appdata and desktop launchers should be part of upstream repository.

**Comment by guihkx** [comment]
bot, build es.estoes.wallpaperDownloader

**Comment by egara** [comment]
Hi all,

@guihkx there are some parameters in  **es.estoes.wallpaperDownloader.sh** that should be added in order to avoid JVM crashes. This should be the full script to run WallpaperDownloader:

```
#!/usr/bin/env bash

exec java -Dappdynamics.force.default.ssl.certificate.validation=false -Dsun.java2d.xrender=f -Xmx256m -Xms128m -jar /app/share/es.estoes.wallpaperDownloader/wallpaperdownloader.jar "$@"
```

Thank you very much :)

Greetings!

**Comment by guihkx** [comment]
Glad you mentioned those, I was going to ask you about them...

Regarding `-Dsun.java2d.xrender=f`:

This one was introduced in [2017](https://bitbucket.org/eloy_garcia_pca/wallpaperdownloader/commits/c3c0f35d), do we still see those `RenderBadPicture` today? Is there a way we can reproduce it? I've used the wallpaper previewer a lot and haven't got any errors, yet.

Regarding `-Dappdynamics.force.default.ssl.certificate.validation=false`:

Correct me if I'm wrong, but only a single person [has reported](https://bitbucket.org/eloy_garcia_pca/wallpaperdownloader/issues/47/self-signed-certificate-problems) these certificate issues, no? Plus, they didn't provide a way to reproduce them. So, I believe that disabling certificate checks is not a very good idea... Thoughts?

**Comment by guihkx** [comment]
We talked by e-mail and decided to keep `-Dsun.java2d.xrender=f` and remove `-Dappdynamics.force.default.ssl.certificate.validation=false`.

Anyway, I wanted to give the wallpaper portal a quick try, but it doesn't seem to be working (P.S. I'm not very familiar with D-Bus). So, I tried adding an override to my existing installation of WallpaperDownloader:

```
$ flatpak --user override --talk-name=org.freedesktop.portal.Wallpaper es.estoes.wallpaperDownloader
```

Then, I tried [listing](https://unix.stackexchange.com/a/46309) the available D-Bus services, but `org.freedesktop.portal.Wallpaper` doesn't show up:

```
$ flatpak run --command=dbus-send es.estoes.wallpaperDownloader --session --dest=org.freedesktop.DBus --type=method_call --print-reply /org/freedesktop/DBus org.freedesktop.DBus.ListNames
method return time=1584081956.001222 sender=org.freedesktop.DBus -> destination=:1.3167 serial=12 reply_serial=2
   array [
      string "org.freedesktop.DBus"
      string "org.freedesktop.Flatpak"
      string "org.freedesktop.portal.Documents"
      string ":1.2329"
      string ":1.2330"
      string ":1.3167"
   ]
```

My idea was to use `dbus-send` afterwards, just to test if the portal would work in all desktop environments (is it supposed to?).

What am I missing? Thanks.

**Comment by barthalion** [comment]
I have no idea either, opening an issue in xdg-desktop-portal repo may be the best choice. I'm fine with merging it as is if you consider it read @guihkx.

**Comment by guihkx** [comment]
Hmm, I might do that later. For now, can you please merge, then? Thank you!

**Comment by barthalion** [comment]
Apparently GitHub has some issues with running actions tonight. I'll re-run merge job tomorrow morning.

**Pull request closed without merging**